Add reader for SQL Databases

kevinqz opened the pull request:
Hey all,

I have:
- Added a reader to connect and query PostgreSQL Databases.
- Included an example (.py file) of how to connect and query a database correctly.
- Updated the docs with the additional PostgreSQL option.

Best,
Kevin Saltarelli

jerryjliu commented:
thanks for doing this @kevinqz! any chance you could take a look at using the SQLDatabase class instead? https://gpt-index.readthedocs.io/en/latest/reference/struct_store.html - this is already used for our SQL index, and we could use it as a reader too (SQLAlchemy should allow for connecting to most structured databases including postgres)

kevinqz commented:
Hi, @jerryjliu! Just added another commit in which I've:

- Added a DatabaseReader to connect to and query SQL Databases by using SQLAlchemy and the SQLDatabase class.
- Included an example (now with a Jupyter Notebook .ipynb file).
- Updated the docs with the additional DatabaseReader option.

Tried a few tests here and everything should be fine to go. Feel free to let me know if anything.

jerryjliu commented on gpt_index/readers/postgresql.py:
```diff
@@ -0,0 +1,57 @@
+"""PostgreSQL client."""
```
can we remove this client if it's superseded by the DatabaseReader?

jerryjliu commented on examples/test_postgres/TestPostgresReader.py:
```diff
@@ -0,0 +1,33 @@
+from __future__ import absolute_import
```
If we're removing the postgres reader, I'd remove this example

jerryjliu commented on examples/test_database.py/TestDatabaseReader.ipynb:
```diff
@@ -0,0 +1,135 @@
+{
```
can we put TestDatabaseReader under the `data_connectors` folder in examples? (and name it DatabaseReaderDemo.ipynb)

jerryjliu commented on gpt_index/readers/database.py:
```diff
@@ -0,0 +1,67 @@
+from gpt_index.readers.schema.base import Document
+from gpt_index.langchain_helpers.sql_wrapper import SQLDatabase
+
+class DatabaseReader(SQLDatabase):
```
oh my original thought was to just pass in the sql_database object as a variable (similar to how we pass in sql_database to the structured index), so the user has to initialize a SQLDatabase object on their own.

I thought about it for a bit, I actually don't mind your current approach too much, but instead I would prefer my proposal for a few reasons: 
- here you're overriding the __init__ signature of SQLDatabase, so the old init signature can't be used
- you should also be inheriting from BaseReader, here you're inheriting from SQLDatabase 
- it's consistent with how users create a structured index atm

With my proposal, `load_data` would do a similar call to sql_database to run the query, return the result, and would then convert them to Document objects

kevinqz commented:
I see your point, @jerryjliu! I've made some changes:

- Made changes in the way to instantiate the DatabaseReader class. Inherited it as a BaseReader and added an instance variable called sql_database of the SQLDatabase class. Please let me know if I understood and implemented your points correctly.
- Also added the ability to start from an existing (SQLAlchemy) Engine - not sure if this is something excessive or necessary, please let me know your opinion.
- Removed 2 old files (related to PostgreSQL only).
- Added examples of how to instantiate the DatabaseReader class from sql_database (SQLDatabase), an engine (Engine), and an uri (string) in the DatabaseReaderDemo.ipynb file.

Feel free to let me know if I correctly addressed your points, ok? Still new to this whole area. :)

jerryjliu commented:
> I see your point, @jerryjliu! I've made some changes:
> 
> * Made changes in the way to instantiate the DatabaseReader class. Inherited it as a BaseReader and added an instance variable called sql_database of the SQLDatabase class. Please let me know if I understood and implemented your points correctly.
> * Also added the ability to start from an existing (SQLAlchemy) Engine - not sure if this is something excessive or necessary, please let me know your opinion.
> * Removed 2 old files (related to PostgreSQL only).
> * Added examples of how to instantiate the DatabaseReader class from sql_database (SQLDatabase), an engine (Engine), and an uri (string) in the DatabaseReaderDemo.ipynb file.
> 
> Feel free to let me know if I correctly addressed your points, ok? Still new to this whole area. :)

nice, this is a lot better! could you remove the old files? there's TestDatabaseReader.ipynb, TestPostgresReader, and postgresql.py which can be removed

kevinqz commented:
@jerryjliu, done! Let me know if I did it correctly this time. 😅
